repair/repair.cc

```diff
@@ -1540,3 +1540,109 @@ future<> sync_data_with_repair(seastar::sharded<database>& db,
+            auto& strat = ks.get_replication_strategy();
+            dht::token_range_vector desired_ranges = strat.get_pending_address_ranges(tm, tokens, myip);
+
+            //Active ranges
+            auto metadata_clone = tm.clone_only_token_map();
+            auto range_addresses = strat.get_range_addresses(metadata_clone);
+
+            //Pending ranges
+            metadata_clone.update_normal_tokens(tokens, myip);
+            auto pending_range_addresses = strat.get_range_addresses(metadata_clone);
+
+            //Collects the source that will have its range moved to the new node
+            std::unordered_map<dht::token_range, repair_neighbors> range_sources;
+
+            rlogger.info("bootstrap_with_repair: started with keyspace={}, nr_ranges={}", keyspace_name, desired_ranges.size());
+            for (auto& desired_range : desired_ranges) {
+                for (auto& x : range_addresses) {
+                    const range<dht::token>& src_range = x.first;
+                    seastar::thread::maybe_yield();
+                    if (src_range.contains(desired_range, dht::tri_compare)) {
+                        std::vector<inet_address> old_endpoints(x.second.begin(), x.second.end());
+                        auto it = pending_range_addresses.find(desired_range);
+                        if (it == pending_range_addresses.end()) {
+                            throw std::runtime_error(format("Can not find desired_range = {} in pending_range_addresses", desired_range));
+                        }
+
+                        std::unordered_set<inet_address> new_endpoints(it->second.begin(), it->second.end());
+                        rlogger.debug("bootstrap_with_repair: keyspace={}, range={}, old_endpoints={}, new_endpoints={}",
+                                keyspace_name, desired_range, old_endpoints, new_endpoints);
+                        // Due to CASSANDRA-5953 we can have a higher RF then we have endpoints.
+                        // So we need to be careful to only be strict when endpoints == RF
+                        // That is we can only have RF >= old_endpoints.size().
+                        // 1) If RF > old_endpoints.size(), it means no node is
+                        // going to lose the ownership of the range, so there
+                        // is no need to choose a mandatory neighbor to sync
+                        // data from.
+                        // 2) If RF = old_endpoints.size(), it means one node is
+                        // going to lose the ownership of the range, we need to
+                        // choose it as the mandatory neighbor to sync data
+                        // from.
+                        std::vector<gms::inet_address> mandatory_neighbors;
+                        // All neighbors
+                        std::vector<inet_address> neighbors;
+                        if (old_endpoints.size() == strat.get_replication_factor()) {
```

Comment:
What about network topology strategy? Don't we want to use the local datacenter (and use only local nodes)?

We can have a per-DC replication factor of 3 and if it is satisfied for the local datacenter but not satisfied for the remote datacenter, then the number of endpoints will be 5 when considered globally (and les than the replication factor of 6) but 3 when considered locally (and the same as the local replication factor).

Reply:
Consider this example, we want to bootstrap a new node in the DC2, 
```
                     DC1_RF=3        DC2_RF=3

1) EXISTING NODES    3                     3
2) EXISTING NODES    2                     2
3) EXISTING NODES    3                     2
4) EXISTING NODES    2                     3
5) EXISTING NODES    3                     0
```

Case 1) is the most common case, it sync with one of the node in DC2. 
Case 2), both DCs do not have enough replicas.   
Case 3) and 4), one of DCs does not have enough replicas. 
Case 5), a new DC is added, no node in the local dc to sync data with. Normally, we won't have this situation because we will set RF = 0 when a new dc is added. 

In current patch set, case 2 to 5 sync with nodes in all DCs. Case 1 will sync with one node in local DC which is DC aware. To sync with less nodes, we can tune case 2 to 4 to sync with local DC only, tune case 5 to sync with one of the remote DC.  Let's decide if we want to do this optimization now. How do you think?

BTW, we have plenty of opportunities to simplify the add new DC procedure with the repair based ops.

Reply:
@avikivity can you take a look at this?

Reply:
I think the algorithm should have two passes: first, check if local datacenter resources are sufficient, and if so use them. If not, consider other datacenters. Or perhaps we should refuse the operation.

For now I'm most concerned with not introducing regressions. How is the patch vs. the existing situation?

Reply:
Before it always streams from one of the existing nodes. After it syncs with one node that is losing the ranges if enough replicas are present (on par with existing situation); if not enough replicas are present, it repairs with more nodes (more work than existing situation). 

How about:

If local dc has enough replicas, repair with the node losing the ranges in local DC.
If local dc does not have enough replicas (but at least one), repair with all replicas nodes in the local DC.
If local dc has zero replica, reject the operation because we do not support this way to add a new DC. We can support this later.

Reply:
s/repair with all/repair with quorum/, and I agree

Reply:
Quorum of what exactly? Quorum of local DC RF or quorum of existing nodes. 

For example:

With RF = 3, we have  1 existing node in the local DC, if we require quorum of RF which is 3/2+1=2, we could not bootstrap the node.

With RF = 5,  we have 3 nodes in the cluster, if we require quorum of existing nodes which is 2 to repair with, this does not guarantee enough RF quorum nodes.

Reply:
> Quorum of what exactly? Quorum of local DC RF or quorum of existing nodes.

LOCAL_QUORUM

> 
> For example:
> 
> With RF = 3, we have 1 existing node in the local DC, if we require quorum of RF which is 3/2+1=2, we could not bootstrap the node.

Good point.

How about this:

If you have a LOCAL_QUORUM, use that
Otherwise, find a LOCAL_QUORUM from another datacenter, and use that
Otherwise, use all existing nodes



> 
> With RF = 5, we have 3 nodes in the cluster, if we require quorum of existing nodes which is 2 to repair with, this does not guarantee enough RF quorum nodes.

Reply:
> If you have a LOCAL_QUORUM, use that
> Otherwise, find a LOCAL_QUORUM from another datacenter, and use that
> Otherwise, use all existing nodes

Let RF = 3, LOCAL_QUORUM = 2, n1 is the only node in local DC, n2 is added. LOCAL_QUORUM can not be achieved when bootstrapping.
With your rule #2 above, bootstrapping node will ignore the local DC node n1, and use 2 nodes from remote nodes instead.  However if the user selects Write CL = 1, Read CL = 3, streaming from n1 is good enough.  It looks to me that using LOCAL_QUORUM to decide which node to sync with assumes the user always use QUORUM. 

I do not want to add the cross DC node selection, since we have the rule "If local dc has zero replica, reject the operation because we do not support this way to add a new DC. We can support this later."

I think we can simplify the rule when LOCAL_QUORUM is not meet case, the idea behind that is in this case we have at most RF / 2 to sync with even if we choose sync with all local dc nodes. 

1) If local dc has RF replicas, sync with the node losing the ranges in local DC, that is one node. 
2) If local dc has RF/2 + 1 replicas, sync with RF/2 + 1 nodes.
3) if local dc has less than RF/2 + 1 replicas, sync with all remaining nodes which is at most RF/2 nodes and at least 1 node.  
4) if local dc has zero replica, reject the operation because we do not support this way to add a new DC. We can support this later.


Note: Rule 3) is safe. For example, we have one node in local dc and  RF=3.  W=1 and R = 3 is the only safe way to read and write if there is only one node, so to sync with the only node in DC is safe. 
Note2: The above can be simplified as 
1)  local_dc_replica_nodes = RF,  sync with 1 node that is losing the range
2)  0 < local_dc_replica_nodes < RF, sync with up to  RF/2 + 1 nodes
3)  local_dc_replica_nodes  = 0, reject